lib/ai_moderation/spam_scanner.rb

```diff
@@ -0,0 +1,349 @@
+          user = User.find_by(id: SiteSetting.ai_spam_detection_user_id)
+        end
+        user || Discourse.system_user
+      end
+
+      def self.after_cooked_post(post)
+        return if !enabled?
+        return if !should_scan_post?(post)
+        return if !post.custom_fields[SHOULD_SCAN_POST_CUSTOM_FIELD]
+        return if post.updated_at < MAX_AGE_TO_SCAN.ago
+
+        last_scan = AiSpamLog.where(post_id: post.id).order(created_at: :desc).first
+
+        if last_scan && last_scan.created_at > EDIT_DELAY_MINUTES.minutes.ago
+          delay_minutes =
+            ((last_scan.created_at + EDIT_DELAY_MINUTES.minutes) - Time.current).to_i / 60
+          Jobs.enqueue_in(delay_minutes.minutes, :ai_spam_scan, post_id: post.id)
+        else
+          Jobs.enqueue(:ai_spam_scan, post_id: post.id)
+        end
+      end
+
+      def self.edited_post(post)
+        return if !enabled?
+        return if !should_scan_post?(post)
+        return if scanned_max_times?(post)
+
+        previous_version = post.revisions.last&.modifications&.dig("raw", 0)
+        current_version = post.raw
+
+        return if !significant_change?(previous_version, current_version)
+
+        flag_post_for_scanning(post)
+      end
+
+      def self.flag_post_for_scanning(post)
+        post.custom_fields[SHOULD_SCAN_POST_CUSTOM_FIELD] = "true"
+        post.save_custom_fields
+      end
+
+      def self.enabled?
+        SiteSetting.ai_spam_detection_enabled && SiteSetting.discourse_ai_enabled
+      end
+
+      def self.should_scan_post?(post)
+        return false if !post.present?
+        return false if post.user.trust_level > TrustLevel[1]
+        return false if post.topic.private_message?
+        if Post
+             .where(user_id: post.user_id)
+             .joins(:topic)
+             .where(topic: { archetype: Archetype.default })
+             .limit(4)
+             .count > 3
+          return false
+        end
+        true
+      end
+
+      def self.scanned_max_times?(post)
+        AiSpamLog.where(post_id: post.id).count >= 3
+      end
+
+      def self.significant_change?(previous_version, current_version)
+        return true if previous_version.nil? # First edit should be scanned
+
+        # Use Discourse's built-in levenshtein implementation
+        distance =
+          ScreenedEmail.levenshtein(previous_version.to_s[0...1000], current_version.to_s[0...1000])
+
+        distance >= MINIMUM_EDIT_DIFFERENCE
+      end
+
+      def self.test_post(post, custom_instructions: nil)
+        settings = AiModerationSetting.spam
+        llm = settings.llm_model.to_llm
+        custom_instructions = custom_instructions || settings.custom_instructions.presence
+        context = build_context(post)
+        prompt = completion_prompt(post, context: context, custom_instructions: custom_instructions)
+
+        result =
+          llm.generate(
+            prompt,
+            temperature: 0.1,
+            max_tokens: 100,
```

Comment:
Same as above

Reply:
sure changed it, and added an extra step in tests for reasoning. the reasoning is helpful